flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,316 @@
+			network,
+			numberOfRemoteChannelsGate2,
+			numberOfLocalChannelsGate2);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+		RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
+
+		SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, tuple2.f0};
+		FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+		CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
+			floatingBuffersUsageGauge,
+			exclusiveBuffersUsageGauge,
+			inputGates);
+
+		try {
+			assertEquals(0.0, exclusiveBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBuffersUsageGauge.getValue(), 0.0);
+
+			int totalBuffers = extraNetworkBuffersPerGate * numberOfInputGates + buffersPerChannel * numberOfRemoteChannelsTotal;
+
+			drainAndValidate(
+				buffersPerChannel,
+				buffersPerChannel,
+				remoteInputChannel1,
+				closeableRegistry,
+				totalBuffers,
+				buffersPerChannel * numberOfRemoteChannelsTotal,
+				exclusiveBuffersUsageGauge,
+				inputBuffersUsageGauge,
+				inputGate1);
+
+			drainAndValidate(
+				buffersPerChannel,
+				buffersPerChannel * 2,
+				remoteInputChannel2,
+				closeableRegistry,
+				totalBuffers,
+				buffersPerChannel * numberOfRemoteChannelsTotal,
+				exclusiveBuffersUsageGauge,
+				inputBuffersUsageGauge,
+				inputGate1);
+		} finally {
+			inputGate1.close();
+			inputGate2.close();
+			closeableRegistry.close();
+			network.close();
+		}
+	}
+
+	@Test
+	public void testFloatingBuffersUsage() throws IOException, InterruptedException {
+
+		int numberOfRemoteChannelsGate1 = 2;
+		int numberOfLocalChannelsGate1 = 0;
+		int numberOfRemoteChannelsGate2 = 1;
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
+		int numberOfInputGates = 2;
+
+		int buffersPerChannel = 2;
+		int extraNetworkBuffersPerGate = 8;
+
+		CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder()
+			.setNetworkBuffersPerChannel(buffersPerChannel)
+			.setFloatingNetworkBuffersPerGate(extraNetworkBuffersPerGate)
+			.build();
+
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate1,
+			numberOfLocalChannelsGate1);
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple2 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate2,
+			numberOfLocalChannelsGate2);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
```

Comment:
this assert is not necessary for metric changes , so we could remove it and then change to tuple2 instead.